cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
"parser is not good anyway" - I mean that my change is just shows the problem, propose a quick workaround (and probably way to fix it) but doesn't pretend on complete solution.